Source/Core/Common/BitFieldView.h

```diff
@@ -0,0 +1,788 @@
+constexpr field_t GetBitFieldFixed(const host_t host)
+{
+  AssertHostTypeIsSane<host_t>();
+  AssertFieldTypeIsSane<field_t>();
+  AssertBitFieldFixedIsSane<field_t, start, width, host_t>();
+
+  using uhost_t = std::make_unsigned_t<host_t>;
+  using shost_t = std::make_signed_t<host_t>;
+  constexpr std::size_t rshift = 8 * sizeof(host_t) - width;
+  constexpr std::size_t lshift = rshift - start;
+
+  // Choose between arithmetic (sign extend) or logical (no sign extend) right-shift.
+  // Fun fact: This is implementaton defined.  It is only possible to do this implicitly.
+  if constexpr (std::is_signed_v<field_t>)
+    return static_cast<field_t>(static_cast<shost_t>(host << lshift) >> rshift);
+  else
+    return static_cast<field_t>(static_cast<uhost_t>(host << lshift) >> rshift);
+}
+
+template <typename field_t, typename host_t>
+constexpr field_t GetBitField(const std::size_t start, const std::size_t width, const host_t host)
+{
+  AssertHostTypeIsSane<host_t>();
+  AssertFieldTypeIsSane<field_t>();
+  AssertBitFieldIsSane<field_t, host_t>(start, width);
+
+  using uhost_t = std::make_unsigned_t<host_t>;
+  using shost_t = std::make_signed_t<host_t>;
+  const std::size_t rshift = 8 * sizeof(host_t) - width;
+  const std::size_t lshift = rshift - start;
+
+  // Choose between arithmetic (sign extend) or logical (no sign extend) right-shift.
+  // Fun fact: This is implementaton defined.  It is only possible to do this implicitly.
+  if constexpr (std::is_signed_v<field_t>)
+    return static_cast<field_t>(static_cast<shost_t>(host << lshift) >> rshift);
+  else
+    return static_cast<field_t>(static_cast<uhost_t>(host << lshift) >> rshift);
+}
+
+template <typename field_t, std::size_t start, std::size_t width, typename host_t>
+constexpr void SetBitFieldFixed(host_t& host, const field_t val)
+{
+  AssertHostTypeIsSane<host_t>();
+  AssertFieldTypeIsSane<field_t>();
+  AssertBitFieldFixedIsSane<field_t, start, width, host_t>();
+
+  using uhost_t = std::make_unsigned_t<host_t>;
+  constexpr std::size_t rshift = 8 * sizeof(host_t) - width;
+  constexpr uhost_t mask = std::numeric_limits<uhost_t>::max() >> rshift << start;
+
+  host = (host & ~mask) | ((static_cast<uhost_t>(val) << start) & mask);
+}
+
+template <typename field_t, typename host_t>
+constexpr void SetBitField(const std::size_t start, const std::size_t width, host_t& host,
+                           const field_t val)
+{
+  AssertHostTypeIsSane<host_t>();
+  AssertFieldTypeIsSane<field_t>();
+  AssertBitFieldIsSane<field_t, host_t>(start, width);
+
+  using uhost_t = std::make_unsigned_t<host_t>;
+  const std::size_t rshift = 8 * sizeof(host_t) - width;
+  const uhost_t mask = std::numeric_limits<uhost_t>::max() >> rshift << start;
+
+  host = (host & ~mask) | ((static_cast<uhost_t>(val) << start) & mask);
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Float host type specialization
```

Comment:
You make the assumption `float` is 32-bit and `double` is 64-bit. This is not always the case. While `Common::BitCast` will check this, you may wish to make this assumption explicit with a `static_assert`.

Reply:
This was what I did in an earlier version of the library, but it got lost in translation.  Thanks.